api/apps/api/src/modules/projects/project-cost-surface.controller.ts

```diff
@@ -0,0 +1,132 @@
+import {
+  GeometryFileInterceptor,
+  GeometryKind,
+} from '@marxan-api/decorators/file-interceptors.decorator';
+import { asyncJobTag } from '@marxan-api/dto/async-job-tag';
+import { RequestWithAuthenticatedUser } from '@marxan-api/app.controller';
+import {
+  AsyncJobDto,
+  JsonApiAsyncJobMeta,
+} from '@marxan-api/dto/async-job.dto';
+import { ensureShapefileHasRequiredFiles } from '@marxan-api/utils/file-uploads.utils';
+import { isLeft } from 'fp-ts/Either';
+import { mapAclDomainToHttpError } from '@marxan-api/utils/acl.utils';
+import { scenarioResource } from '@marxan-api/modules/scenarios/scenario.api.entity';
+import { ScenariosService } from '@marxan-api/modules/scenarios/scenarios.service';
+import { CostRangeDto } from '@marxan-api/modules/scenarios/dto/cost-range.dto';
+import { plainToClass } from 'class-transformer';
+import { Response } from 'express';
+import { marxanRunFiles } from '@marxan-api/modules/scenarios/scenarios.controller';
+import { JwtAuthGuard } from '@marxan-api/guards/jwt-auth.guard';
+import { ImplementsAcl } from '@marxan-api/decorators/acl.decorator';
+
+@ApiTags(projectResource.className)
+@Controller(`${apiGlobalPrefixes.v1}/projects`)
+export class ProjectCostSurfaceController {
+  constructor(
+    @Inject(forwardRef(() => ScenariosService))
+    public readonly scenarioService: ScenariosService,
+  ) {}
+
+  @ImplementsAcl()
+  @UseGuards(JwtAuthGuard)
+  @ApiConsumesShapefile({ withGeoJsonResponse: false })
+  @GeometryFileInterceptor(GeometryKind.ComplexWithProperties)
+  @ApiTags(asyncJobTag)
+  @Post(`:scenarioId/cost-surface/shapefile`)
+  async processCostSurfaceShapefile(
+    @Param('scenarioId') scenarioId: string,
+    @Req() req: RequestWithAuthenticatedUser,
+    @UploadedFile() file: Express.Multer.File,
+  ): Promise<JsonApiAsyncJobMeta> {
+    await ensureShapefileHasRequiredFiles(file);
+
+    const result = await this.scenarioService.processCostSurfaceShapefile(
+      scenarioId,
+      req.user.id,
+      file,
+    );
+
+    if (isLeft(result)) {
+      throw mapAclDomainToHttpError(result.left, {
+        scenarioId,
+        userId: req.user.id,
+        resourceType: scenarioResource.name.plural,
+      });
+    }
+    return AsyncJobDto.forScenario().asJsonApiMetadata();
+  }
+
+  @ImplementsAcl()
+  @UseGuards(JwtAuthGuard)
+  @Get(`:scenarioId/cost-surface`)
+  @ApiOkResponse({ type: CostRangeDto })
+  async getCostRange(
+    @Param('scenarioId') scenarioId: string,
+    @Req() req: RequestWithAuthenticatedUser,
+  ): Promise<CostRangeDto> {
+    const result = await this.scenarioService.getCostRange(
+      scenarioId,
+      req.user.id,
+    );
+    if (isLeft(result)) {
+      throw mapAclDomainToHttpError(result.left, {
+        scenarioId,
+        userId: req.user.id,
+        resourceType: scenarioResource.name.plural,
+      });
+    }
+    return plainToClass<CostRangeDto, CostRangeDto>(CostRangeDto, result.right);
+  }
```

Comment:
we'll likely end up changing this - the URL was kinda wonky all along (it should be something like `@Get(':scenarioId/cost-surface/cost-range')`) but I think in the id26.2 implementation we could _store_ min and max for each cost surface at upload/creation time (in `(apidb)cost_surfaces`) and just return that, no need to do even an index scan (I hope it is currently an index scan, not a table scan, I haven't checked right now the indexes we may rely on).

Frontend will need the `[min, max]` range both for individual cost surfaces (we use this to set the gradient colours) and for the active cost surface for each scenario, which could be done transparently in the API by using the same code path we'd use to retrieve the range of a cost surface, for the cost surface that is active in the given scenario.

TL;DR we'll need a "cost-range" endpoint both in this `ProjectCostSurfaceController` and in the `ScenariosController`, for the two use cases in the previous sentence 😓

Reply:
Again, I totally missed the big picture. The id26.2 you mention covers a lot of stuff that I am yet to fully understand. I am not sure how to proceed with this specific suggestion.